examples/evaluator/golang/simple/evaluate/evaluator.go

```diff
@@ -15,40 +15,78 @@
 package evaluate
 
 import (
-	"open-match.dev/open-match/examples"
+	"math"
+	"sort"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/sirupsen/logrus"
 	harness "open-match.dev/open-match/pkg/harness/evaluator/golang"
 	"open-match.dev/open-match/pkg/pb"
 )
 
 // Evaluate is where your custom evaluation logic lives.
 // This sample evaluator sorts and deduplicates the input matches.
 func Evaluate(p *harness.EvaluatorParams) ([]*pb.Match, error) {
-	scoreInDescendingOrder := func(a, b *pb.Match) bool {
-		return a.GetProperties().GetFields()[examples.MatchScore].GetNumberValue() > b.GetProperties().GetFields()[examples.MatchScore].GetNumberValue()
+	matches := make(matcheExts, 0, len(p.Matches))
+	for _, match := range p.Matches {
+		// Evaluation criteria is optional, but sort it lower than any matches which
+		// provided criteria.
+		ext := &pb.DefaultEvaluationCriteria{
+			Score: math.Inf(-1),
+		}
+		if match.Extension != nil {
+			err := ptypes.UnmarshalAny(match.Extension, ext)
+			if err != nil {
+				p.Logger.WithFields(logrus.Fields{
+					"match_id": match.MatchId,
+					"error":    err,
+				}).Error("Failed to unmarshal match's DefaultEvaluationCriteria.  Rejecting match.")
+				continue
+			}
+		}
+		matches = append(matches, &matchExt{
+			match: match,
+			ext:   ext,
+		})
 	}
-	by(scoreInDescendingOrder).Sort(p.Matches)
+
+	sort.Sort(matches)
 
 	results := []*pb.Match{}
-	dedup := map[string]bool{}
+	ticketsUsed := map[string]struct{}{}
 
-	for _, match := range p.Matches {
-		if isNonCollidingMatch(match, dedup) {
-			for _, ticket := range match.GetTickets() {
-				dedup[ticket.GetId()] = true
+outer:
+	for _, matchExt := range matches {
+		for _, ticket := range matchExt.match.GetTickets() {
+			if _, ok := ticketsUsed[ticket.Id]; ok {
+				continue outer
```

Comment:
Can we use a helper function to check hasDuplicateTickets? I tend to agree with the discussion and solution here for that being a relatively elegant (and readable) solution for nested loops

Reply:
What discussion?

I think keeping the test on ticketsUsed and the update on it relatively close improves this code.  I could move it all to another function, but that requires more structure just to have a return statement instead of a continue.

Reply:
Sorry - forgot to link the Stackoverflow thread i was referring to: https://stackoverflow.com/questions/51996175/how-to-break-out-of-nested-loops-in-go

Specifically referring to the use of labels to control flow - I do think helper function that returns / breaks is more readable.

Reply:
for _, matchExt := range matches {
    if containsUsedTickets(match, ticketsUsed) {
       continue;
    }
    // append
}

Reply:
done.